### PR TITLE
Remove all global log invocations

### LINKED ISSUE #21: Don't use global logger; take a logger as a parameter

### PR DESCRIPTION
**Pull request by peterbourgon**
Fixes #21

### CODE REVIEW

File: connection.go [hunk 1]
```diff
@@ -30,15 +30,17 @@ type remoteConnection struct {
 	remoteTCPAddr string
 	outbound      bool
 	established   bool
+	logger        *log.Logger
```
Comment (rade) [inline connection.go]: This should not need a logger. See #33.

**Comment by rade** [comment]
Is putting a `logger` field into most structs, and adding an extra arg to the initialisation methods really the best way to go about this?

Most things have access to `router.Ourself`, so could just use the logger in that.

Though ironically with my changes in #33 and the changes in here, I think LocalPeer isn't actually doing any logging itself anymore - it only needs a logger in order to feed it into the LocalConnection constructor.

**Comment by peterbourgon** [comment]
I elected to make the minimum changes necessary. I am happy to refactor further. May I delete methods like remoteConnection.log, or is that prefix useful/necessary?

**Comment by rade** [comment]
#33 gets rid of remoteConnection.log.

File: connection.go (hunk 1 above)
Comment (peterbourgon) [inline connection.go]: Don't see how to avoid it, given remoteConnection implements log method.

File: gossip_test.go [hunk 2]
```diff
@@ -22,7 +24,7 @@ var _ gossipConnection = &mockGossipConnection{}
 
 func newTestRouter(name string) *Router {
 	peerName, _ := PeerNameFromString(name)
-	router := NewRouter(Config{}, peerName, "nick", nil)
+	router := NewRouter(Config{}, peerName, "nick", nil, log.New(ioutil.Discard, "", 0))
```
Comment (rade) [inline gossip_test.go]: really? That might make debugging test failures rather hard.

File: peers.go [hunk 3]
```diff
@@ -549,7 +552,7 @@ func decodePeer(dec *gob.Decoder) (ps peerSummary, connSummaries []connectionSum
 	return
 }
 
-func makeConnsMap(peer *Peer, connSummaries []connectionSummary, byName map[PeerName]*Peer) map[PeerName]Connection {
+func makeConnsMap(peer *Peer, connSummaries []connectionSummary, byName map[PeerName]*Peer, logger *log.Logger) map[PeerName]Connection {
```
Comment (rade) [inline peers.go]: logger is no longer used here. And hence I think you can get rid of it from `Peer` altogether.

File: gossip_test.go (hunk 2 above)
Comment (peterbourgon) [inline gossip_test.go]: I am optimizing for quiet logs in the happy path. In the case of failures this has worked well for me in the past:

1. Test fails
2. Modify test, add logger
3. Run test again
4. Identify failure

Is that insufficient here?

Comment (rade) [inline gossip_test.go]: @bboreham has been debugging many test failures in this area, so I'd like to get his take.

Comment (bboreham) [inline gossip_test.go]: I agree with @peterbourgon's general thrust, but would add that at step 2 one typically wants a logger with debug-level turned on.

Maybe a package-scope variable for the logger would be better than a locally-created one you have to edit?

I can't think of any cases where we were able to debug a rare test failure specifically from the non-debug logs.  Maybe from the panic trace - that reminds me we might want to override Go 1.6's change to printing just the current goroutine.

Comment (peterbourgon) [inline gossip_test.go]: - There is no leveled logging; all logs are debug logs.
- This PR specifically eliminates package global loggers.

Given both of these things, I'll take that as a tepid approval. Thanks for the commentary.

**Comment by bboreham** [comment]
It would be easier to integrate this with other schemes if it took an interface rather than a concrete `*log.Logger`.

**Comment by peterbourgon** [comment]
Agreed. https://github.com/weaveworks/mesh/issues/37

**Pull request merged**